node/engine/messageservice/p2p-message-service/dht-record.go

```diff
@@ -0,0 +1,85 @@
+package p2pms
+
+import (
+	"encoding/json"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+const (
+	DHT_RECORD_PREFIX = "/scaddr/"
+)
+
+type stateChannelAddrToPeerIDValidator struct{}
+
+// DhtRecord represents the data stored in the DHT record
+type DhtRecord struct {
+	Data      DhtData `json:"data"`
+	Signature []byte  `json:"signature"`
+}
+
+type DhtData struct {
+	SCAddr    string `json:"scaddr"` // state channel address
+	PeerID    string `json:"peerid"`
+	Timestamp int64  `json:"timestamp"` // Unix timestamp (seconds since January 1, 1970)
+}
+
+func (v stateChannelAddrToPeerIDValidator) Validate(key string, value []byte) error {
+	// Trim the DHT_RECORD_PREFIX from the key to get the state channel address
+	signingAddrStr := strings.TrimPrefix(key, DHT_RECORD_PREFIX)
+
+	// Check if it's a valid state channel address
+	if !common.IsHexAddress(signingAddrStr) {
+		return errors.New("invalid state channel address used for key")
+	}
+
+	// Parse the value into a RecordData object
+	var dhtRecord DhtRecord
+	if err := json.Unmarshal(value, &dhtRecord); err != nil {
+		return errors.New("malformed record value")
+	}
+
+	// Make sure the timestamp is not in the future or negative number
+	if dhtRecord.Data.Timestamp > time.Time.Unix(time.Now()) || dhtRecord.Data.Timestamp < 0 {
+		return errors.New("invalid timestamp")
+	}
+
+	// Check if the value can be parsed into a valid libp2p peer.ID
+	peerId, err := peer.Decode(dhtRecord.Data.PeerID)
+	if err != nil {
+		return errors.New("invalid libp2p peer ID")
+	}
+
+	pubKey, err := peerId.ExtractPublicKey()
+	if err != nil {
+		return err
+	}
+
+	dataBytes, err := json.Marshal(dhtRecord.Data)
+	if err != nil {
+		return err
+	}
+
+	// Check the signature to ensure it is the signed hash of dataBytes
+	valid, err := pubKey.Verify(dataBytes, dhtRecord.Signature)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		return errors.New("invalid signature")
+	}
+
+	return nil
+}
+
+// Simply return the first record as the best record.
+// In a more complex scenario, we could add logic to select the best record
+// based on some criteria.
+func (v stateChannelAddrToPeerIDValidator) Select(key string, values [][]byte) (int, error) {
```

Comment:
Should we return an error in this function if there are multiple entries(IE: `len(values)>1`) for now? I think we only expect one record to be published per SC address, so if there are multiple records something has gone wrong.

Reply:
The `Select` method is used to choose between records if different ones are returned by different peers for the same key, which is possible because the dht is not updated instantly across all nodes so some peers can have stale data before a record update is fully propagated. I don't think it should cause an error if we receive two different values when we query for a given key. Rather, I think we should choose the record that has the latest timestamp to ensure we have the most recently updated entry.

Reply:
I added logic to the `Select` method to choose the most recent record we receive

Reply:
> Rather, I think we should choose the record that has the latest timestamp to ensure we have the most recently updated entry.

Should we make that change in this PR? I'm just worried that always returning the first record might introduce some subtle race conditions that could be difficult to debug.

Reply:
Yes it has been added